tests/unit/test_results.py

```diff
@@ -41,6 +45,64 @@ def _get_vcr(self, **kwargs):
             "Set-Cookie",
             "User-Agent",
         ]
+        myvcr.filter_query_parameters = [
+            ("client_id", REDACTED_STRING),
+        ]
+
+        def redact_key_values(keys_to_redact):
+            def before_record_response(response):
+                # Only do this if the response has not been recorded
+                string_body = response["body"]["string"].decode("utf8")
+                if REDACTED_STRING not in string_body:
+                    # Only do this is if the body contains one or more
+                    # of the keys to redact.
+                    if any(key in string_body for key in keys_to_redact):
+                        try:
+                            is_list = False
+                            # Marshall into json object, if it is a JSON object.
+                            payload = json.loads(string_body)
+                            if isinstance(payload, list):
+                                payload = payload[0]
+                                is_list = True
+                            for key in keys_to_redact:
+                                if key in payload:
+                                    # Redact the key value
+                                    payload[key] = REDACTED_STRING
+                            # Write out the updated json object to the response
+                            # body string.
+                            if is_list:
+                                payload = [payload]
+                            response["body"]["string"] = json.dumps(payload).encode()
+                        except ValueError:
+                            # If it is not a json object, return
+                            return response
+
+                return response
```

Comment:
To simplify this logic a bit, I suggest _not_ bothering to avoid redacting an already redacted payload. This redaction is idempotent, so redacting an already redacted payload yields the same result, and in the context of running unit tests, the performance impact is likely unnoticeable.

```suggestion
            def redact(payload):
                for key in keys_to_redact:
                    if key in payload:
                        payload[key] = "REDACTED"
                return payload

            def before_record_response(response):
                body = response["body"]["string"].decode("utf8")

                with contextlib.suppress(json.JSONDecodeError):
                    payload = json.loads(body)
                    redacted_payload = map(redact, payload) if isinstance(payload, list) else redact(payload)
                    response["body"]["string"] = json.dumps(redacted_payload).encode()

                return response
```

Also, please move the entirety of `redact_key_values` to the top level. It does not need to be nested with the `_get_vcr` method. The nesting just adds unnecessary clutter.

Same goes for the nested `redact_login_request` below.

Reply:
Oddly, I move this method out of TestCase it no longer gets called. Investigating.